wasm/src/lsp/parse.rs

```diff
@@ -126,3 +144,660 @@ fn test_doc_vec_byte_iter_0() {
+pub fn is_identifier(v: &[u8]) -> bool {
+    v.iter().copied().all(is_identifier_char)
+}
+
+#[test]
+fn test_is_identifier() {
+    assert!(is_identifier(b"test"));
+}
+
+#[test]
+fn test_is_not_identifier() {
+    assert!(!is_identifier(b"; comment instead"));
+}
+
+fn make_inner_function_scopes(scopes: &mut Vec<ParseScope>, body: &BodyForm) {
+    match body {
+        BodyForm::Let(LetFormKind::Sequential, letdata) => {
+            if letdata.bindings.is_empty() {
+                make_inner_function_scopes(scopes, letdata.body.borrow());
+                return;
+            }
+
+            let binding = &letdata.bindings[0];
+            let new_location = if letdata.bindings.len() == 1 {
+                letdata.body.loc()
+            } else {
+                letdata.bindings[1].loc().ext(&letdata.loc.ending())
+            };
+
+            let mut variables = HashSet::new();
+            variables.insert(SExp::Atom(binding.nl.clone(), binding.name.clone()));
+
+            let mut inner_scopes = Vec::new();
+
+            make_inner_function_scopes(
+                &mut inner_scopes,
+                &BodyForm::Let(
+                    LetFormKind::Sequential,
+                    LetData {
+                        loc: new_location.clone(),
+                        kw: letdata.kw.clone(),
+                        bindings: letdata.bindings.iter().skip(1).cloned().collect(),
+                        body: letdata.body.clone(),
+                    },
+                ),
+            );
+
+            scopes.push(ParseScope {
+                region: new_location,
+                kind: ScopeKind::Let,
+                variables,
+                functions: HashSet::new(),
+                containing: inner_scopes,
+            });
+        }
+        BodyForm::Let(LetFormKind::Parallel, letdata) => {
+            if letdata.bindings.is_empty() {
+                make_inner_function_scopes(scopes, letdata.body.borrow());
+                return;
+            }
+
+            let mut name_set = HashSet::new();
+            for b in letdata.bindings.iter() {
+                let new_name = SExp::Atom(b.nl.clone(), b.name.clone());
+                name_set.insert(new_name);
+            }
+
+            let mut inner_scopes = Vec::new();
+            make_inner_function_scopes(&mut inner_scopes, letdata.body.borrow());
+
+            let new_scope = ParseScope {
+                region: letdata.bindings[0].loc.ext(&letdata.body.loc().ending()),
+                kind: ScopeKind::Let,
+                variables: name_set,
+                functions: HashSet::new(),
+                containing: inner_scopes,
+            };
+            scopes.push(new_scope);
+        }
+        BodyForm::Call(_, v) => {
+            for elt in v.iter() {
+                make_inner_function_scopes(scopes, elt);
+            }
+        }
+        _ => {}
+    }
+}
+
+fn make_helper_scope(h: &HelperForm) -> Option<ParseScope> {
```

Comment:
`cargo build` warned that this function is never used.
This should be made public or removed.